### PR TITLE
Topic recognition: Siamese model classifier ADNI dataset

### PR DESCRIPTION
**Pull request by MTbir**
Name: Minh Tieu
Student Number: 46419712

Task 7: Create a classifier based on Siamese network to classify Alzheimer’s disease (normal and AD) of the ADNI brain data set having an accuracy around 0.8 on the test set. [Hard Difficulty]

This pull request include the the implementation of the OneShot Siamese network that will take an image and return 4096 feature vectors. After the Siamese network is finished training, the feature vector will be run through a binary classifier (5 fully connected classifier) to classify whether the image is Alzheimer's or disease.

Instruction on how to reproduce the run is inside README.md and other detail.

To run the code, there are two file you should run. If the model is train and generate siamese.pt and classifier.pt, you can run ```python3 predict.py``` to generate prediction on a random batch of 12 images. The image will include the predict classification and the real label. In addition, it will also return test accuracy result. If siamese.py or classifier.py don't exist, run ```python3 train.py``` and everything will be generate. Please download the dataset yourself and update utils.py for the save path as well as path to look at the train data. The final accuracy is around 77%

**Commit File and Folder**
**Result folder**
The folder include all the image related to the accuracy of the training and testing. For example, the accuracy of use the classifier on validating data, classifier loss, siamese loss, t-SNE diagram about feature vector on train dataset, validate dataset and test dataset (only the last epoch is include) and the classify image compare between the predict vs actual label.

**Research image**
The folder include all the theory based research from online source that help explain Siamese model theory.

README.md - This is the report of this project. It include everything you need to know about this project, model, implementation, result and how to reproduce the data. All images in both result and research is included inside this.

```utils.py```: This include the path that use to load/store the model or plot.
```dataset.py```: This include all relevant information about dataloader, typically, data augmentation, custom data preprocessing for splitting train data to 80% and 20% validating (Patient-wise to prevent data leakage) and Pair data loader for Siamese train.
```modules.py```: This file include Siamese model convolution layer, Binary classifier fully connected layer and Contrastive loss for Siamese model. Everything related to construct the model.
```train.py```: This file include all the training, this is where the both model is train. The Siamese is finished training first then it will be used to extract the feature vector to train the binary classifier. After finish training, it will also print out the test accuracy to stdout but also save loss model for siamese and  classifier as well as save siamese and classifier parameters in siamese.pt and classifier.pt. In addition, for the purpose of testing, the main function from ```predict.py``` is also include in ```train.py``` so, when running ```train.py```, it is the same as running ```train.py``` then running ```predict.py```. For more information, look in README.md.
```predict.py```: This file load the model and print out the test accuracy in stdout and also save the image about the classification of the test data predict label vs real label.
```slurm.sh```: Script for rangpur job -> showing what I run when training.

There are a total of 15 file inclusive

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Very good
- clear flow of information
- very detailed information on algorithm, training choices, hyperparameter tuning, usage
- relevant plots and figures shown
- results followed with good discussion/interpretations

Commit messages: Good

Code: 
- good design
- sufficiently commented
- algorithm seem to be correctly implemented

Functionality/Performance: 
- both siamese and classifier overfit massively
- classifier achieves 77% accuracy on test set

General comments:
- algorithm solves the problem apporpriately with slightly below-expected performance

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
